src/main/scala/com/microsoft/hyperspace/index/rules/JoinIndexRuleV2.scala

```diff
@@ -0,0 +1,221 @@
+    // All references for any side, either left or right, should all come from the same leaf node.
+    //
+    // E.g. if join condition refers to columns A, B, C, D, two columns should come from left, 2
+    // from right. This would be by default true. Nothing to do in this part. Let's say A,B
+    // come from left. C,D come from right. The requirement is both A and B should come from the
+    // same leaf node on left. Same for C and D. Both should come from same leaf node on right.
+    //
+    // TODO: Improve this algorithm based on whether number of buckets for an index match
+    //  "spark.sql.shuffle.partitions". Refer https://github.com/microsoft/hyperspace/issues/237.
+    require(condition.references.forall(_.isInstanceOf[AttributeReference]))
+    val joinCols =
+      extractConditions(condition).flatMap(_.children.map(_.asInstanceOf[AttributeReference]))
+    val eligibleBaseRelations = plan.collectLeaves().filter {
+      case r: LogicalRelation if !isPlanModified(r) => r.output.exists(contains(joinCols, _))
+      case _ => false
+    }
+
+    eligibleBaseRelations match {
+      case Seq(r: LogicalRelation)
+          if r.output.toSet.count(c => contains(joinCols, c)) * 2 == joinCols.size =>
+        updateIndex(plan, joinCols, plan.output)
+      case _ => plan
+    }
+  }
+
+  /**
+   * Update the plan by replacing source data with indexes. To make sure the index covers all
+   * required columns, traverse down the plan and collect all referenced columns from all nodes
+   * starting from top of the plan to the bottom.
+   *
+   * Please note: we can't use plan.transform(...) here because we have to collect referenced
+   * columns from every node we traverse till the end.
+   * @param plan Original plan.
+   * @param joinCols All referenced join columns from join condition.
+   * @param requiredCols Superset of all columns which are referenced in the plan. If an index
+   *                     is chosen to replace a base relation `r`, it should contain all the
+   *                     columns from this `requiredCols`, which directly belong to `r`.
+   * @return Logical plan with indexes if applicable indexes are found, else the original plan.
+   */
+  private def updateIndex(
+      plan: LogicalPlan,
+      joinCols: Seq[AttributeReference],
+      requiredCols: Seq[Attribute]): LogicalPlan = {
+    plan match {
+      case p: LogicalRelation => setIndexes(p, joinCols, requiredCols)
+      case p: LeafNode => p
+      case p: LogicalPlan =>
+        p.withNewChildren {
+          p.children.map { child =>
+            updateIndex(child, joinCols, p.references.toSeq ++ requiredCols)
+          }
+        }
+    }
+  }
+
+  private def eligible(l: LogicalPlan, r: LogicalPlan, condition: Expression): Boolean = {
+    !isBroadcastJoin(l, r) && isJoinConditionSupported(condition)
+  }
+
+  private def isBroadcastJoin(l: LogicalPlan, r: LogicalPlan): Boolean = {
```

Comment:
Seems this doesn't work as expected.

Logical Plan (tpcds q91):
```
+- 'Aggregate ['cc_call_center_id, 'cc_name, 'cc_manager, 'cd_marital_status, 'cd_education_status], ['cc_call_center_id AS Call_Center#4158, 'cc_name AS Call_Center_Name#4159, 'cc_manager AS Manager#4160, 'sum('cr_net_loss) AS Returns_Loss#4161]
   +- 'Filter ((((('cr_call_center_sk = 'cc_call_center_sk) && ('cr_returned_date_sk = 'd_date_sk)) && ('cr_returning_customer_sk = 'c_customer_sk)) && ((('cd_demo_sk = 'c_current_cdemo_sk) && ('hd_demo_sk = 'c_current_hdemo_sk)) && ('ca_address_sk = 'c_current_addr_sk))) && (((('d_year = 1998) && ('d_moy = 11)) && ((('cd_marital_status = M) && ('cd_education_status = Unknown)) || (('cd_marital_status = W) && ('cd_education_status = Advanced Degree)))) && ('hd_buy_potential LIKE Unknown% && ('ca_gmt_offset = -7))))
      +- 'Join Inner
         :- 'Join Inner
         :  :- 'Join Inner
         :  :  :- 'Join Inner
         :  :  :  :- 'Join Inner
         :  :  :  :  :- 'Join Inner
         :  :  :  :  :  :- 'UnresolvedRelation `call_center`
         :  :  :  :  :  +- 'UnresolvedRelation `catalog_returns`
         :  :  :  :  +- 'UnresolvedRelation `date_dim`
         :  :  :  +- 'UnresolvedRelation `customer`
         :  :  +- 'UnresolvedRelation `customer_address`
         :  +- 'UnresolvedRelation `customer_demographics`
         +- 'UnresolvedRelation `household_demographics`
```

physical plan:
```
== Physical Plan ==
*(11) Sort [Returns_Loss#4161 DESC NULLS LAST], true, 0
+- Exchange rangepartitioning(Returns_Loss#4161 DESC NULLS LAST, 200)
   +- *(10) HashAggregate(keys=[cc_call_center_id#6, cc_name#11, cc_manager#16, cd_marital_status#235, cd_education_status#236], functions=[sum(cr_net_loss#111)], output=[Call_Center#4158, Call_Center_Name#4159, Manager#4160, Returns_Loss#4161])
      +- Exchange hashpartitioning(cc_call_center_id#6, cc_name#11, cc_manager#16, cd_marital_status#235, cd_education_status#236, 200)
         +- *(9) HashAggregate(keys=[cc_call_center_id#6, cc_name#11, cc_manager#16, cd_marital_status#235, cd_education_status#236], functions=[partial_sum(cr_net_loss#111)], output=[cc_call_center_id#6, cc_name#11, cc_manager#16, cd_marital_status#235, cd_education_status#236, sum#4174])
            +- *(9) Project [cc_call_center_id#6, cc_name#11, cc_manager#16, cr_net_loss#111, cd_marital_status#235, cd_education_status#236]
               +- *(9) BroadcastHashJoin [c_current_hdemo_sk#254L], [hd_demo_sk#343L], Inner, BuildRight
                  :- *(9) Project [cc_call_center_id#6, cc_name#11, cc_manager#16, cr_net_loss#111, c_current_hdemo_sk#254L, cd_marital_status#235, cd_education_status#236]
                  :  +- *(9) BroadcastHashJoin [c_current_cdemo_sk#253L], [cd_demo_sk#233L], Inner, BuildRight
                  :     :- *(9) Project [cc_call_center_id#6, cc_name#11, cc_manager#16, cr_net_loss#111, c_current_cdemo_sk#253L, c_current_hdemo_sk#254L]
                  :     :  +- *(9) BroadcastHashJoin [c_current_addr_sk#255L], [ca_address_sk#207L], Inner, BuildRight
                  :     :     :- *(9) Project [cc_call_center_id#6, cc_name#11, cc_manager#16, cr_net_loss#111, c_current_cdemo_sk#253L, c_current_hdemo_sk#254L, c_current_addr_sk#255L]
                  :     :     :  +- *(9) SortMergeJoin [cr_returning_customer_sk#92L], [c_customer_sk#251L], Inner
                  :     :     :     :- *(4) Sort [cr_returning_customer_sk#92L ASC NULLS FIRST], false, 0
                  :     :     :     :  +- Exchange hashpartitioning(cr_returning_customer_sk#92L, 200)
                  :     :     :     :     +- *(3) Project [cc_call_center_id#6, cc_name#11, cc_manager#16, cr_returning_customer_sk#92L, cr_net_loss#111]
                  :     :     :     :        +- *(3) BroadcastHashJoin [cr_returned_date_sk#85L], [d_date_sk#287L], Inner, BuildRight
                  :     :     :     :           :- *(3) Project [cc_call_center_id#6, cc_name#11, cc_manager#16, cr_returned_date_sk#85L, cr_returning_customer_sk#92L, cr_net_loss#111]
                  :     :     :     :           :  +- *(3) BroadcastHashJoin [cc_call_center_sk#5L], [cr_call_center_sk#96L], Inner, BuildLeft
                  :     :     :     :           :     :- BroadcastExchange HashedRelationBroadcastMode(List(input[0, bigint, true]))
                  :     :     :     :           :     :  +- *(1) Project [cc_call_center_sk#5L, cc_call_center_id#6, cc_name#11, cc_manager#16]
                  :     :     :     :           :     :     +- *(1) Filter isnotnull(cc_call_center_sk#5L)
                  :     :     :     :           :     :        +- *(1) FileScan parquet [cc_call_center_sk#5L,cc_call_center_id#6,cc_name#11,cc_manager#16] Batched: true, Format: Parquet, Location: InMemoryFileIndex[wasb://teststorage@testtest.blob.core.windows.net/data/tpcds-1T..., PartitionFilters: [], PushedFilters: [IsNotNull(cc_call_center_sk)], ReadSchema: struct<cc_call_center_sk:bigint,cc_call_center_id:string,cc_name:string,cc_manager:string>                  :     :     :     :           :     +- *(3) Project [cr_returned_date_sk#85L, cr_returning_customer_sk#92L, cr_call_center_sk#96L, cr_net_loss#111]
                  :     :     :     :           :        +- *(3) Filter ((isnotnull(cr_call_center_sk#96L) && isnotnull(cr_returned_date_sk#85L)) && isnotnull(cr_returning_customer_sk#92L))
                  :     :     :     :           :           +- *(3) FileScan parquet [cr_returned_date_sk#85L,cr_returning_customer_sk#92L,cr_call_center_sk#96L,cr_net_loss#111] Batched: true, Format: Parquet, Location: InMemoryFileIndex[wasb://teststorage@testtest.blob.core.windows.net/data/tpcds-1T..., PartitionFilters: [], PushedFilters: [IsNotNull(cr_call_center_sk), IsNotNull(cr_returned_date_sk), IsNotNull(cr_returning_customer_sk)], ReadSchema: struct<cr_returned_date_sk:bigint,cr_returning_customer_sk:bigint,cr_call_center_sk:bigint,cr_net...
                  :     :     :     :           +- BroadcastExchange HashedRelationBroadcastMode(List(input[0, bigint, true]))
                  :     :     :     :              +- *(2) Project [d_date_sk#287L]
                  :     :     :     :                 +- *(2) Filter ((((isnotnull(d_year#293) && isnotnull(d_moy#295)) && (d_year#293 = 1998)) && (d_moy#295 = 11)) && isnotnull(d_date_sk#287L))
                  :     :     :     :                    +- *(2) FileScan parquet [d_year#293,d_date_sk#287L,d_moy#295] Batched: true, Format: Parquet, Location: InMemoryFileIndex[wasb://teststorage@testtest.blob.core.windows.net/indexes/tpcds..., PartitionFilters: [], PushedFilters: [IsNotNull(d_year), IsNotNull(d_moy), EqualTo(d_year,1998), EqualTo(d_moy,11), IsNotNull(d_date_sk)], ReadSchema: struct<d_year:int,d_date_sk:bigint,d_moy:int>
                  :     :     :     +- *(5) Project [c_customer_sk#251L, c_current_cdemo_sk#253L, c_current_hdemo_sk#254L, c_current_addr_sk#255L]
                  :     :     :        +- *(5) Filter (((isnotnull(c_customer_sk#251L) && isnotnull(c_current_addr_sk#255L)) && isnotnull(c_current_cdemo_sk#253L)) && isnotnull(c_current_hdemo_sk#254L))
                  :     :     :           +- *(5) FileScan parquet [c_customer_sk#251L,c_current_addr_sk#255L,c_current_cdemo_sk#253L,c_current_hdemo_sk#254L] Batched: true, Format: Parquet, Location: InMemoryFileIndex[wasb://teststorage@testtest.blob.core.windows.net/indexes/tpcds..., PartitionFilters: [], PushedFilters: [IsNotNull(c_customer_sk), IsNotNull(c_current_addr_sk), IsNotNull(c_current_cdemo_sk), IsNotNull..., ReadSchema: struct<c_customer_sk:bigint,c_current_addr_sk:bigint,c_current_cdemo_sk:bigint,c_current_hdemo_sk..., SelectedBucketsCount: 200 out of 200
                  :     :     +- BroadcastExchange HashedRelationBroadcastMode(List(input[0, bigint, true]))
                  :     :        +- *(6) Project [ca_address_sk#207L]
                  :     :           +- *(6) Filter ((isnotnull(ca_gmt_offset#218) && (ca_gmt_offset#218 = -7.0)) && isnotnull(ca_address_sk#207L))
                  :     :              +- *(6) FileScan parquet [ca_address_sk#207L,ca_gmt_offset#218] Batched: true, Format: Parquet, Location: InMemoryFileIndex[wasb://teststorage@testtest.blob.core.windows.net/indexes/tpcds..., PartitionFilters: [], PushedFilters: [IsNotNull(ca_gmt_offset), EqualTo(ca_gmt_offset,-7.0), IsNotNull(ca_address_sk)], ReadSchema: struct<ca_address_sk:bigint,ca_gmt_offset:double>
                  :     +- BroadcastExchange HashedRelationBroadcastMode(List(input[0, bigint, true]))
                  :        +- *(7) Project [cd_demo_sk#233L, cd_marital_status#235, cd_education_status#236]
                  :           +- *(7) Filter ((((cd_marital_status#235 = M) && (cd_education_status#236 = Unknown)) || ((cd_marital_status#235 = W) && (cd_education_status#236 = Advanced Degree))) && isnotnull(cd_demo_sk#233L))
                  :              +- *(7) FileScan parquet [cd_demo_sk#233L,cd_education_status#236,cd_marital_status#235] Batched: true, Format: Parquet, Location: InMemoryFileIndex[wasb://teststorage@testtest.blob.core.windows.net/indexes/tpcds..., PartitionFilters: [], PushedFilters: [Or(And(EqualTo(cd_marital_status,M),EqualTo(cd_education_status,Unknown)),And(EqualTo(cd_marital..., ReadSchema: struct<cd_demo_sk:bigint,cd_education_status:string,cd_marital_status:string>
                  +- BroadcastExchange HashedRelationBroadcastMode(List(input[0, bigint, true]))
                     +- *(8) Project [hd_demo_sk#343L]
                        +- *(8) Filter ((isnotnull(hd_buy_potential#345) && StartsWith(hd_buy_potential#345, Unknown)) && isnotnull(hd_demo_sk#343L))
                           +- *(8) FileScan parquet [hd_demo_sk#343L,hd_buy_potential#345] Batched: true, Format: Parquet, Location: InMemoryFileIndex[wasb://teststorage@testtest.blob.core.windows.net/indexes/tpcds..., PartitionFilters: [], PushedFilters: [IsNotNull(hd_buy_potential), StringStartsWith(hd_buy_potential,Unknown), IsNotNull(hd_demo_sk)], ReadSchema: struct<hd_demo_sk:bigint,hd_buy_potential:string>
```

It might be because of planLater of lower level of BHJ?

Reply:
@imback82 @apoorvedave1 I added physical plan of disabled hyperspace & query execution/explain time.
I cannot see any difference of the plan except for removed 1 Exchange.

Maybe it's because of explain time, but no idea why it takes longer.
In any case, this shows `isBroadcastJoin` function doesn't work properly.

Other queries also show regressions, e.g.:
Result - hyperspace enabled (only ran 1 query) 
```
***** Query Execution Times *****
q21             18933   15805   15275

***** Query Explain Times *****
q21             6218    1554    1730
```
Result - disabled hyperspace
```
***** Query Execution Times *****
q21             13479   10073   8741

***** Query Explain Times *****
q21             483     359     225
```
I'll share the plans for q21 if needed.

I think we need to make sure there's no regression from this rule, at least for tpcds testset, non-hybrid scan.
@apoorvedave1 Could you measure the perf and add required constraints to avoid these regressions? Thanks!